### ISSUE TITLE
code completion: add operator suggestions

### ISSUE THREAD

**Issue by cruzdanilo** [clangd/clangd#603]
add suggestions and auto complete for operators as member functions, for example:
```c++
const auto u = name_map->operator[](u_name);
```
obs: this feature is available on `ms-vscode.cpptools`

**Comment by HighCommander4** [comment]
(Migrated issue to clangd repo as this is a server-side feature request.)

**Comment by sam-mccall** [comment]
Trying to understand: 
1. you want to type `->oper` and get `->operator[]` inserted?
2. Or you want to type `->oper` and get `[]` inserted?

I believe currently we filter out operators because 1 is the most natural behavior but it produces unnatural code: explicitly calling operators with that syntax is quite rare. (Anecdotally, I've seen `(*name_map)[u_name]` much more often, ugly as it is).

2 might be interesting from an API discovery point of view, but it's surprising and certainly doesn't save any typing.

**Comment by HighCommander4** [comment]
I understood it as (1). I think that's useful, especially with pointers, as the `(*map)[name]` syntax is more cumbersome to write.

**Comment by cruzdanilo** [comment]
actually, i would like to type `name_map->` and see a list with all the possibilities, which is exactly how i learned about this syntax in the past, while using `ms-vscode.cpptools`. but it would also save typing if you use this kind of syntax.

about the syntax, i use operators as member functions for code consistency, using `->` like everywhere else. before i knew about this syntax, i used to prefer `map->at(key)`, even though it's less performant because of bound checking. the other option, `(*name_map)[u_name]` is ugly and different from every other use of a pointer.

just as a reference, `ms-vscode.cpptools` shows a nice distinct icon for operators:
![Screen Shot 2020-11-26 at 19 43 41](https://user-images.githubusercontent.com/216636/100384164-14d6b900-3052-11eb-8805-00c435513685.png)

on a side note, `ms-vscode.cpptools` also highlights overloaded operators (using either syntax) with the same color as any other member function (yellowish on the default theme). i wonder if the lack of operator highlighting in `vscode-clangd` is related.
![Screen Shot 2020-11-26 at 19 47 10](https://user-images.githubusercontent.com/216636/100384402-aba37580-3052-11eb-8ac4-ae2393c3d8e2.png)

**Comment by sam-mccall** [comment]
I'm concerned these might be spammy. We could limit this by any of:
 - only providing them after `->`, not `.`
 - only supporting a set of operators where this syntax is likely useful (would anyone write `foo->operator=(bar)` instead of `*foo=bar`? or `operator*`?)
 - only providing them for user-declared operators (i.e. specifically not *implicit* operator=)

What do you think? I get the argument that it's nice if code completion tells you everything that's legal in the context, but that's pretty far from where we are now, and doing that consistently produces a lot of mostly-irrelevant noise (destructors, overridden base members, etc)

**Comment by cruzdanilo** [comment]
>- only providing them after `->`, not `.`

i don't think this makes sense, as code completion replaces `.` with `->` when necessary and i often use this feature because it's easier and shorter.

>- only providing them for user-declared operators (i.e. specifically not _implicit_ operator=)

i like that.

>- only supporting a set of operators where this syntax is likely useful (would anyone write `foo->operator=(bar)` instead of `*foo=bar`? or `operator*`?)

because of the last one, i think it would be nice to know of any overloaded operators, even `foo->operator=(bar)`, especially when documentation is available.
